### DISCUSSION TITLE
CMakePresets

### DISCUSSION THREAD

**Discussion by L4stR1t3s** [o3de/o3de#2363] [Ideas]
We already discussed CMakePresets on the Discord, but I thought we could use a single place to discuss and decide on what the Open3D Engine CMakePresets should be.

More information on CMakePresets: https://cmake.org/cmake/help/latest/manual/cmake-presets.7.html

TL;DR: CMakePresets.json is used to configure build options and is used by IDEs like Visual Studio and Visual Studio Code (ones I know off). CMakeUserPresets.json can be used to create personal build options.

What needs to be done/decided?

1. Add CMakeUserPresets.json to .gitignore.
2. Create a CMakePresets.json in the root folder.
3. Define a number of presets in CMakePreset.json based on: Platforms? Configurations? Targets? Unity? Something else?
4. Decide what needs to be defined in CMakePresets.json: package directory, build directory, install directory, unity build, ... Generally we should provide a default value for the values set on the command line when running cmake.

We can also use this to solve issues like https://github.com/o3de/o3de/issues/2275. By adding an empty toolchain file and linking to it in CMAKE_TOOLCHAIN_FILE in CMakePresets.json for Visual Studio, we can stop Visual Studio from injecting the vcpkg path, potentially causing CMake to use the vcpkg packages instead of the downloaded packages.

I would suggest that we don't try to fill in every possible preset that would logically need to exist. I think it's probably best if someone starts a CMakePresets.json and creates some for the platform they are working on. I can start it off for Windows if there are no objections. Others can add presets for their platform of choice if they want to use presets. I think we may not need any for platforms that don't have an IDE that uses presets.

So I would make a presets for:
- Windows
- Debug/Release/Profile
- Targets (Can someone who knows what targets exist provide a list?)
- Unity or not.

I would suggest the following naming conventions:
- platform (3 characters): win, mac, lin, ...
- configuration (3 characters): dbg, rel, prf
- target: full name
- unity: optional

Separated by "-".
Platform and configuration are shortened, so the target is more visible.

Examples:
win-dbg-editor
win-rel-editor-unity
mac-prf-testers-unity

For the values of the package and install folder I would suggest using the default values that are set when not adding them. On Windows they are put in the current users folder, which should have write access. I don't know what other platforms do. For the build directory, "build" in the root.

What did I forget? Do wrong? ...

**Comment by amznestebanpapp** [comment]
Thanks for putting this into a discussion.

We currently have a set of settings that are used by Jenkins to do build validation. Those settings look like could be presets and have these Jenkins scripts just use presets. 
The settings are per platform, for e.g. Windows they are here: `scripts\build\Platform\Windows\build_config.json`

> So I would make a presets for:
> 
> Windows
> Debug/Release/Profile
> Targets (Can someone who knows what targets exist provide a list?)
> Unity or not.

For targets, I suggest to start with:
- all
- Editor

Game/Server targets are tricky because the name of the target contains the name of the project, so we have to figure that one out (e.g. maybe there is some variable that we can pass).

> platform (3 characters): win, mac, lin, ...

For Platforms is better to follow the naming convention we have in PAL (Platform abstraction layers) since a lot of the scripts already understand that: Windows, Mac, iOS, Linux, Android, etc. You can see these names in a lot of "Platform" folders, also around in the source code and in PAL files.
Those names are also already handled by other scripts like `ci_build.py`, is also consistent with e.g. parameters we present in `Jenkinsfile`

> configuration (3 characters): dbg, rel, prf

Same for configuration, since you could define your own configuration, we should follow the same naming convention: "debug", "profile", "release".

> target: full name

Target will get a bit complicated because we likely will define "groups". For the time being seems the two groups are "all" and "Editor"

> unity: optional

Should be unity be put at the end? Note that there are options that can be parametrized so the user has to enable it or not.
Another option should be "monolithic", usually someone would do a "release monolithic" to reduce space and trim everything that is not being used in a game.

> Separated by "-".

Is there any special separator in IDEs?
https://cmake.org/cmake/help/latest/manual/cmake-presets.7.html seems to also separate using "-". Some UIs sometimes organize things if you name it with a "/" separator (e.g. I name my git branches that way because UIs organize it in folders). 
If there is no special separator handling, then looks like "-" is the way to go (VS and cmake documentation use that)

> Platform and configuration are shortened, so the target is more visible.

Is that a UI issue? the combobox in VS is usually resizable. Dont want it to get out control but some shortening could make it harder for a developer to match things across scripts and conventions (e.g. shortening configuration and platform).

We should also try to reuse and reduce as much as possible these presets. There is some "inheritance" so we can handle that.
There are two more things that eventually will add more:
- architecture: once we support e.g. ARM on linux (or potentially in Windows)
- generator: e.g. if you want to generate Ninja in Windows or VS2022 in Windows, or Makefiles in Linux. Ideally we only want to provide options for one generator per platform to reduce the amount of options that we need to validate and support. If we put a generator that is not validated (in Jenkins) a user/dev may assume that it is supported. However, for Windows, we may need to support 2 generators e.g. once 2022 becomes official, we may need to validate VS2019 and VS2022.
I am also open to put the generator name there (or a shorter name) so if someone were to add other generators in a user preset they can distinguish them.

**Reply by L4stR1t3s** [reply]
3 chars: Because then the target names line up nicer in a list, but it's not such a big deal and I agree that using the same naming convention everywhere is more important.

Unity: I would make a unity and non-unity preset for each platform/configuration/target so people can decide what they would like to build.

"-": I think that's the standard way of doing it. Visual Studio creates "x64-debug" as a default if no presets are found for example. I have never seen "/" used for this.

Inheritance: I will creating "hidden" presets so the actual presets will just need for example "inherit": ["windows", "debug","all", "unity", ...]

architecture: perhaps i should add x64 already in the names and make a preset for it? For example: "windows-x64-debug-all". If there will be different architecture, we would have to add it later anyway.

Generator: I would argue that for these presets "Ninja" is a more logical choice on Windows, since it's generally used by people who want to use the CMake/Ninja build flow. Visual Studio uses Ninja as the default generator for this very reason. If someone wants to use solutions/projects they should not open the folder the folder in Visual Studio first, but run cmake or cmake-gui first.

I was planning to ask at some point why other generators are considered "not supported"? That seems contradictory to what CMake is supposed to do: translate a generic description of a project to any build system.

We can add the generator to the name as well: "windows-x64-debug-ninja-all"

**Reply by lemonade-dm** [reply]
FYI CMake now supports a Ninja-Multi-Config generator, that creates all the configurations.
That remove the need to specify the configuration as part of the `configurePresets`

**Reply by lemonade-dm** [reply]
Also the restrictions on using "non-Visual Studio" generators on Windows as been recently relaxed by @burelc-amzn in commit https://github.com/o3de/o3de/commit/da02d47069a4c0492b45310e9236c41d7755b8bc

For other non Visual Studio Generators it is not that they are "not-supported", but more that they are not "officially-supported".
This means that Jenkins builds and testing workflows don't configure with other generators(Ninja, NMake, etc...), but they may work.
The engine makes no guarantees that the other generators work on Windows work, but the engine no longer goes out of it's way to prevent users from attempting to use them.

**Reply by L4stR1t3s** [reply]
> 
> 
> FYI CMake now supports a Ninja-Multi-Config generator, that creates all the configurations.
> That remove the need to specify the configuration as part of the `configurePresets`

The problem is that you have to set the binary dir on the configuration presets level. Debug/release/profile is set on the build presets level, where you can't define the binary dir. The CMakePresets assumes that debug/release/profile are sub folders of the binary directory, but the Open3D Engine CMake scripts don't do that.

So if I would use the "Ninja Multi-Config" generator, it would mean that debug/release/profile would all wind up in the same folder, which we definitely want to avoid.

Unless the CMake scripts are changed so nested folders are created, I don't see another way then creating stub debug/release/profile configuration presets. And  the "Ninja Multi-Config" generator can't be used either, because the different configuration presets need to be set.

**Comment by lemonade-dm** [comment]
With CMake Presets, the user still has the options to override variables on the command line, so that can be used as option for specfiying the install directory via CMAKE_INSTALL_PREFIX.

The relevant targets I see as important enough to add to the preset files are `Editor` and `AssetProcessor"` for building the Tools applications. For generating an install layout the `INSTALL` target is needed. Finally for building the Test Targets and running them, the `TEST_SUITE_smoke` and  `TEST_SUITE_main` targets should be added as an entry to the CMakePresets.json.

I also think we would should add a CMakePresets.json file to the Project Templates located in the [engine-root/Templates](https://github.com/o3de/o3de/tree/development/Templates/DefaultProject/Template) directory, so that any newly created project has the presets available for use in a project-centric workflow where CMake is run from the project directory.

Furthermore the project is what creates the `${Project}.GameLauncher` and `${Project}.ServerLauncher` target, so it would be useful to have preset for generating the Game/Server Launcher for the project itself. We should also update the .gitignore file within the Project Templates themselves to also exclude the `CMakeUserPresets.json` file

**Reply by L4stR1t3s** [reply]
I was planning to look at the project generator and do the same there after this is done.

**Comment by L4stR1t3s** [comment]
An initial version: [CMakePresets.txt](https://github.com/o3de/o3de/files/6865550/CMakePresets.txt) (txt because github wouldn't accept json)

The way this works is that you define configuration, build and test presets. Build and test presets depend on a specific configuration preset.

I had to "hack" the configurations (debug, release, profile) a bit. Since the Open3D Engine CMake scripts don't create nested build folders (windows/x64/debug), I had to create debug, release and profile in the configuration presets as well, to be able to set a different binary directory.

What does this look like in Visual Studio?

Configurations:
![configs](https://user-images.githubusercontent.com/69721170/126718174-5d08572a-6d5f-404f-90c1-46a4e774f120.jpg)

Targets:
![targets](https://user-images.githubusercontent.com/69721170/126718177-e2c92e82-1201-440c-978a-930d55bab78c.jpg)

Ugly names? Yes, "name" has to be a unique field and Visual Studio 2019 does not use the "displayName" field unfortunately. It should look like this:

Configurations:
Windows x64 Debug Ninja
Windows x64 Debug Ninja Unity
Windows x64 Profile Ninja
Windows x64 Profile Ninja Unity
Windows x64 Release Ninja
Windows x64 Release Ninja Unity

Targets:
All
Editor
Asset Processor
Install

Visual Studio 2022 does use "displayName" (https://developercommunity.visualstudio.com/t/cmakepresets-shows-name-in-the-selection-drop-down/1414245) so I wouldn't really worry about it. It's a pretty new feature so tools are still implementing the full scope of it.



Architecture is just a stub for now, as there isn't really anything to set.

I still need to add tests, but that's in a separate section, and I first want to get some feedback on this before I finish it all.

I also still need to add monolithic, but I wasn't sure what exactly to do with it. What exactly do I need to set to make it a monolithic configuration? How many should I add? 1 for each configuration that already exists? Or a subset?

I will add a visual studio generator as well.

And I also want to take a look at conditions, which can be used to filter the available presets. For example: Add a condition on the "windows" configuration preset to only show it on a windows platform.

And finally, I think it might be worth writing a script that generates this file from a script. It's easy to represent what is in the file with a lot less data, as a lot of it is repeated in the CMakePresets json. I would still put the CMakePresets.json in the project, but when it needs to be changed you should use the script and matching data file. That way it would be possible to simply add a single line for a new target and have the script generate all the necessary platform/architecture/configuration/generator/unity/... entries for it. As those should be the same for all the targets, right?

**Reply by lemonade-dm** [reply]
For a monolithic build, what changes is the cmake configuration. The `LY_MONOLITHIC_GAME=1` turns on monolithic builds and makes it so that Client Gems now build into a static library that is linked into the GameLauncher, such that no shared libraries are loaded when running.
Monolithic builds are normally for the final release of the game. So what current is specified in the build_config.json for monolithic is only a monolithic release combination.

Also for the non-host platforms of Android and iOS, there are "Tools" targets, so there is no need to make a target entry for "Asset Processor" or "Editor".

Also FYI when the Visual Studio Generator is used, the target to build "all" becomes "ALL_BUILD".

**Reply by L4stR1t3s** [reply]
> Also for the non-host platforms of Android and iOS, there are "Tools" targets, so there is no need to make a target entry for "Asset Processor" or "Editor".

That's only for Android and iOS Right? I will leave creating presets up to someone familiar with those platforms anyway, because I have zero experience developing for those platforms.

> Also FYI when the Visual Studio Generator is used, the target to build "all" becomes "ALL_BUILD".

Could someone with solutions/projects generated by visual studio try building the "all" target from the command line with cmake? I think CMake considers "all" and "ALL_BUILD" aliases.

**Reply by lemonade-dm** [reply]
Other members will add any presets for iOS and Android as well we need them, so that can be deferred.



> Could someone with solutions/projects generated by visual studio try building the "all" target from the command line with cmake? I think CMake considers "all" and "ALL_BUILD" alias

CMake does not conssider "all" and "ALL_BUILD" an alias
```
D:\o3de\o3de>cmake --build build\windows_vs2019 --target all --config profile
Microsoft (R) Build Engine version 16.10.2+857e5a733 for .NET Framework
Copyright (C) Microsoft Corporation. All rights reserved.

MSBUILD : error MSB1009: Project file does not exist.
Switch: all.vcxproj
```

One thing about invoking the build command though CMake, is that if a target isn't supplied it build all of them.
So explicitly having an "all" target isn't needed.

**Reply by L4stR1t3s** [reply]
> CMake does not conssider "all" and "ALL_BUILD" an alias

Then it must be Visual Studio that is interpreting all as ALL_BUILD, because if I use "all" as a target in the presets, and let Visual Studio call CMake to configure (so not on the command line), it works. I'll change it because it has to work in all situations.

**Reply by amznestebanpapp** [reply]
> > CMake does not conssider "all" and "ALL_BUILD" an alias
> 
> Then it must be Visual Studio that is interpreting all as ALL_BUILD, because if I use "all" as a target in the presets, and let Visual Studio call CMake to configure (so not on the command line), it works. I'll change it because it has to work in all situations.

all is just valid in some generators. It is actually up to that generator. Ninja/Makefiles/etc have an "all", but generators like VS/XCode dont, so CMake puts a "ALL_BUILD" to mimic that. 
I think in all cases, not passing any target will actually build all.

**Reply by amznestebanpapp** [reply]
> > Also for the non-host platforms of Android and iOS, there are "Tools" targets, so there is no need to make a target entry for "Asset Processor" or "Editor".
> 
> That's only for Android and iOS Right? I will leave creating presets up to someone familiar with those platforms anyway, because I have zero experience developing for those platforms.

It also applies to consoles. Basically, any platform that is a "target" but not a "host" for editing tools will fall into this case.
That is why I suggest to have "Tools" instead of "Editor/AP". The AssetProcessor preset is also not very usable for developers. This is needed for Jenkins to run asset jobs, in that case we are interested in the AssetProcessorBatch.
"Tools" will also include things like the MaterialEditor and other applications.

I think we should just define a subset of the "variants" we have + all, the variants we have are: Builders, Tools, Clients, Servers. So I think we should define: Tools, Clients, Servers, All. Builders could be what we use for the asset jobs.
Tools -> build editor and dependencies. For something like Android, this will build for Windows (or Mac) with required tools enabled for the target platform.
Clients -> builds game (this will be generated by the template for the project-centric workflow)
Servers -> builds game server (this will be generated by the template for the project-centric workflow)
Builders -> builds APBatch and dependencies (mainly to be used by Jenkins, can be added later)
All -> builds all

**Comment by L4stR1t3s** [comment]
@lumberyard-employee-dm
@amznestebanpapp

Proposal for generating CMakePresets: https://github.com/L4stR1t3s/o3de_cmakepresets

Why generate? CMakePresets.json gets really big, really  fast, making it difficult to read and edit. Most of the data can be simplified for the specific needs of Open3D Engine and generated, making it easier to read and edit.

I would suggest that whoever needs to make a change, edits the simplified file and generates the CMakePresets.json file. Rather than making this an additional step everyone has to execute, because it's non standard and once CMakePresets are defined, they are unlikely to change often anyway.

This is just a prototype, so file names, data and such are not final.

**Reply by lemonade-dm** [reply]
I think it is a good idea to generate the cmake presets as well.
It makes it easier to extend with newer platforms or configurations.

If we do generate file though, it would be ideal to not have the CMakePresets.json be part of source control.
But it wouldn't be a good customer experience to tell them to generate the file on first pull of the o3de repo.

The trade-off with checking the CMakePresets.json file into source control, is that users would just start to directly modify it to add new presets, instead of running the script again.

It would be good if `git` had some kind of post merge hook that was able to run a script after a specific file was modified.
Such as if the json file that was used to generate the presets was modified, then on the next pull of that change, the script would run automatically and update the CMakePresets.json file.
Though that does have issue as well. The script could fail or the user is using a different source control system.

So I am thinking the safest approach is to have the CMakePresets.json as part of the o3de repo along with the script and template that can generate it as well

**Reply by L4stR1t3s** [reply]
The CMakePresets.json (source controlled) typically should not be edited by a user to make local changes. That's what CMakeUserPresets.json (not source controlled) is for.

The script could be used to generate a CMakeUserPresets.json file as well. We could maybe also add a few example data files in the future for users to start with.

I guess the next step would be to finalize a first version of the presets and script.

I still want to make the script take arguments for input and output file, rather than the hardcoded paths now.

And some of the hardcoded options right now might be better if they were added to the source data file as well (CMakePresets version and minimum required CMake version, as well as a hardcoded path, I forget what it was).

My current understanding is to have the following definitions for a first version:

- platforms: windows (other platforms to be added later by people who use the platform rather than me trying to guess some of the appropriate values)
- architectures: x64 (in the future ARM for example could be added)
- generators: msvc2019, ninja
- configurations: debug, debug-unity, release, release-unity, release-monolithic, release-unity-monolithic, profile, profile-unity
- targets: ALL_BUILD (msvc2019 only), all (ninja only), Editor, Asset Processor, INSTALL

We would also need to decide where to put the script and data file(s). I would like to keep them together, unless there are objections to that? Is "/scripts/cmakepresets/" acceptable?

I guess a bat and/or sh file that calls the script and updates the CMakePresets.json in the root folder with the source controled default data file might also be nice to have.

**Reply by lemonade-dm** [reply]
By modifications, I mean when a user wants t oadd a new engine preset, such as if they wanted to add the generator of `msvc2022`.
Most likely their first step would be to edit the CMakePresets.json file instead of editing the template within the scripts folder and re-run the generation script.

> We would also need to decide where to put the script and data file(s). I would like to keep them together, unless there are objections to that? Is "/scripts/cmakepresets/" acceptable?

Yeah that is a good location. It is a script and it generates CMakePresets. Exactly what it says on the tin.


A helper batch and shell script in the same "scripts/cmakepresets" directory is fine. It would act similer to the o3de.bat/o3de.sh scripts that wrap the python/python.cmd and python/python.sh invocations.

**Reply by L4stR1t3s** [reply]
Oh, I forget the tests (main and smoke). They're already in the presets I defined.

**Reply by L4stR1t3s** [reply]
> Most likely their first step would be to edit the CMakePresets.json file instead of editing the template within the scripts folder and re-run the generation script.

Yea, it's a bit unfortunate that we can't just add a comment at the top of the json file stating that the file must not be edited, but generated instead with steps on how to do it. At best we could add a field that doesn't exist in the CMakePresets schema, but you run the risk of the file being rejected of course by whatever is parsing it.

I guess it's something that will have to be caught by the code review process.

Or is there some way git could be configured to give an error if a file is pushed without another file in the same commit? Then maybe the CMakePresets.json could only be submitted along with the base presets file.

**Reply by amznestebanpapp** [reply]
TBH, I think your `CMakePresets.txt` wasn't that long or unmanageable.  If you consider all the `build_config.json` that we have and do a similar things, this file looks acceptable IMO.
I would suggest to start with your `CMakePresets.json` and we can later decide if it is worthy or not to generate, we can always go back and do it. I think if we have `CMakePresets.json`  we can start using some things like macros and conditions for other platforms and iterate over it before taking a more complicated road.

**Reply by L4stR1t3s** [reply]
@amznestebanpapp
The CMakePresets.txt was just a starting point, that only contained a few presets, not all.
My locally generated CMakePresets.json for just windows platform is currently 1190 lines long. The input file used to generate it is 76 lines long.
Adding a platform that would reuse the existing architectures, generators, configurations and targets would add 1 extra line, and extend a few lines elsewhere in the input json file. It would also almost double the lines in the generated CMakePresets.json.

**Reply by amznestebanpapp** [reply]
sounds good then to go with the generation road.

The execution of the script can become part of the bootstrap process, potentially even called by the installer. 

We can also add a validation rule that runs the script and checks that there are no changes in CMakePresets.json, that way we fail if someone touched the file or if someone modified the script/input file and forgot to generate it.
Check `scripts\commit_validation\commit_validation\validators` those are validations that are run in CI
